server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -147,149 +147,208 @@ public static IndexRequest getIndexWriteRequest(DocWriteRequest docWriteRequest)
+        private final BulkRequest bulkRequest;
+        private final Task task;
+        long startTime = -1;
+
+        BulkExecutor(Task task, BulkRequest bulkRequest, ActionListener<BulkResponse> listener) {
+            super(listener);
+            this.recoveredObserver = new ClusterStateObserver(clusterService, bulkRequest.timeout(), logger, threadPool.getThreadContext());
+            this.bulkRequest = bulkRequest;
+            this.task = task;
+        }
+
+        @Override
+        protected void doRun() {
+            if (startTime == -1) {
+               startTime = relativeTime();
+            }
+            ClusterBlockException blockException = clusterService.state().blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
+            if (blockException != null) {
+                if (recoveredObserver.isTimedOut()) {
+                    // we running as a last attempt after a timeout has happened. don't retry
+                    listener.onFailure(blockException);
+                    return;
+                }
+                recoveredObserver.waitForNextChange(new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState state) {
+                        run();
                     }
-                    // check the alias for the action request (this is how upserts are modeled)
-                    if (indexMetaData == null && actionRequest.index() != null) {
-                        AliasOrIndex indexOrAlias = metaData.getAliasAndIndexLookup().get(actionRequest.index());
-                        if (indexOrAlias != null && indexOrAlias.isAlias()) {
-                            AliasOrIndex.Alias alias = (AliasOrIndex.Alias) indexOrAlias;
-                            indexMetaData = alias.getWriteIndex();
-                        }
+
+                    @Override
+                    public void onClusterServiceClose() {
+                        listener.onFailure(new NodeClosedException(clusterService.localNode()));
                     }
-                    if (indexMetaData != null) {
-                        // Find the default pipeline if one is defined from and existing index.
-                        String defaultPipeline = IndexSettings.DEFAULT_PIPELINE.get(indexMetaData.getSettings());
-                        indexRequest.setPipeline(defaultPipeline);
-                        if (IngestService.NOOP_PIPELINE_NAME.equals(defaultPipeline) == false) {
-                            hasIndexRequestsWithPipelines = true;
+
+                    @Override
+                    public void onTimeout(TimeValue timeout) {
+                        // Try one more time...
+                        run();
+                    }
+                });
+                return;
+            }
+
+            // All good, begin preparing for the bulk request
+            prepForBulk();
+        }
+
+        private void prepForBulk() {
```

Comment:
I think we should avoid getting a fresh cluster state in this method and instead use the one we just verified to be free of any unwanted blocks, i.e. that was passed into `onNewClusterState`.